### PR TITLE
feat(gatekeeper): add chain command parsing in Layer 1

### PR DESCRIPTION
**Pull request by amondnet**
## Summary

- Add quote-aware chain command parsing in Layer 1 of the gatekeeper plugin so chained commands like `npm test && npm run build` can be evaluated without an AI round-trip
- Implement `splitChainedCommands()`, `evaluateSingleCommand()`, and `hasUnquotedChainOps()` helper functions
- Update `evaluate()` to route chained commands through per-part evaluation with conservative security defaults

### New Functions

- **`splitChainedCommands(cmd)`**: Quote-aware state machine parser that splits on `&&`, `||`, `;`, and `|` operators while correctly handling single-quoted, double-quoted, and backtick-quoted segments
- **`evaluateSingleCommand(cmd, rules)`**: Evaluates a single (non-chained) command against the existing DENY/ALLOW rule lists, returning `approve`, `deny`, or `passthrough`
- **`hasUnquotedChainOps(cmd)`**: Detects whether a command string contains unquoted chain operators to distinguish single commands from chains

### Security Invariants Preserved

| Scenario | Behavior |
|---|---|
| `$()` or backtick substitution | Always `passthrough` |
| Newline (`\n`) in command | Always `passthrough` |
| Any part of chain is `deny` | Whole chain is `deny` |
| Any part of chain is `passthrough` | Whole chain is `passthrough` (conservative) |
| Malformed chain (empty parts, unclosed quotes) | `passthrough` |
| Single command (no chain ops) | Identical behavior to before |
| All parts `approve` | Chain is `approve` |

## Test Plan

- [x] 158 unit tests passing covering new functions and updated integration tests
- [x] `splitChainedCommands()` tested for: simple operators, quoted strings with operators inside, nested quotes, empty parts, unclosed quotes
- [x] `evaluateSingleCommand()` tested against all existing DENY/ALLOW rule patterns
- [x] `hasUnquotedChainOps()` tested for operator detection edge cases
- [x] `evaluate()` integration tests updated to verify chain routing and security invariants
- [x] Single-command path verified to produce identical results to pre-change behavior

---
## Summary by cubic
Adds a unified, quote-aware chain parser to Gatekeeper Layer 1 so safe ; and && chains (e.g., npm test && npm run build) are evaluated locally and auto-approved when each part is safe. Complex or malformed commands passthrough; any denied part blocks the chain.

- **New Features**
  - Introduced parseChainedCommand; splitChainedCommands now splits only ; and && outside quotes.
  - Added evaluateSingleCommand; evaluate routes chains per-part (deny on any deny, passthrough on any unknown, allow only if all allow) and generates a composite allow reason.

- **Bug Fixes**
  - Trimmed command input in evaluate to prevent leading-whitespace DENY bypass.
  - Pipes (|), OR (||), redirects (<, >, >>), process substitution (<(), >()), subshell/backticks, newline, lone &, and trailing backslash now passthrough.
  - Hardened DENY rules: block inline interpreter execution (-e/-c/--eval/-p/--print) and find -exec/-execdir/-delete; correctly allow npx -p/--package.

<sup>Written for commit a23fdd7df528b4e098b50fdc1a3d1641c844925e. Summary will update on new commits.</sup>

### CODE REVIEW

File: plugins/gatekeeper/src/pre-tool-use.ts [hunk 1]
```diff
@@ -19,6 +22,16 @@ export const DENY_RULES: Rule[] = [
     pattern: /^dd\s+if=\/dev\/zero\s+of=\/dev\//i,
     reason: 'Disk zeroing blocked',
   },
+  // Block inline code execution in interpreter commands
+  {
+    pattern: /^(node|npx|tsx|python3?|ruby|perl)\s+(-e|-c|--eval)\b/i,
```
[the remark this answers is not in the document]
Comment (amondnet) [reply plugins/gatekeeper/src/pre-tool-use.ts]: ✅ Applied in [3ef548b](https://github.com/pleaseai/claude-code-plugins/commit/3ef548b)

- File: `plugins/gatekeeper/src/pre-tool-use.ts`
- Change: Extended interpreter DENY rule to also block `-p` and `--print` flags: `/^(node|npx|tsx|python3?|ruby|perl)\s+(-e|-p|-c|--eval|--print)\b/i`

Thanks for catching this, @cubic-dev-ai!

File: plugins/gatekeeper/src/chain-parser.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+      }
+      if (ch === '"') {
+        state = 'normal'
+      }
+      current += ch
+    }
+  }
+
+  // Unclosed quote → malformed
+  if (state !== 'normal') {
```
[the remark this answers is not in the document]
Comment (amondnet) [reply plugins/gatekeeper/src/chain-parser.ts]: ✅ Applied in [3ef548b](https://github.com/pleaseai/claude-code-plugins/commit/3ef548b)

- File: `plugins/gatekeeper/src/chain-parser.ts`
- Changes:
  1. Extended post-loop malformed check: `if (state !== 'normal' || escaped)` to detect trailing backslash
  2. Updated quick exit regex from `/[;&|<>]/` to `/[;&|<>\\]/` so commands with backslashes enter the state machine rather than bypassing it via the early return

Thanks for the detailed analysis, @cubic-dev-ai!

File: plugins/gatekeeper/src/pre-tool-use.ts [hunk 3]
```diff
@@ -19,6 +22,16 @@ export const DENY_RULES: Rule[] = [
     pattern: /^dd\s+if=\/dev\/zero\s+of=\/dev\//i,
     reason: 'Disk zeroing blocked',
   },
+  // Block inline code execution in interpreter commands
+  {
+    pattern: /^(node|npx|tsx|python3?|ruby|perl)\s+(-e|-p|-c|--eval|--print)\b/i,
```
[the remark this answers is not in the document]
Comment (amondnet) [reply plugins/gatekeeper/src/pre-tool-use.ts]: ✅ 적용 완료 ([a23fdd7](https://github.com/pleaseai/claude-code-plugins/commit/a23fdd7))

유효한 이슈입니다. `npx -p`는 `--package` (패키지 설치) 플래그로 `node -p`의 `--print` (코드 평가)와 다릅니다.

- 파일: `plugins/gatekeeper/src/pre-tool-use.ts`
- 변경: 인라인 인터프리터 DENY 규칙을 분리
  - `node|tsx`: `-p` 포함 차단 (`--print`이 코드를 평가)
  - `npx|python3?|ruby|perl`: `-p` 제외 (`npx -p`는 `--package`)
- 테스트 추가: `npx -p create-react-app my-app`가 올바르게 허용됨을 검증

제안 감사드립니다 @cubic-dev-ai!
